### ISSUE TITLE
Rails 7 Support

### ISSUE THREAD

**Issue by dankimio** [bwillis/versioncake#108]
Rails 7.0.0.rc1

```
$ bin/rails app:update --trace
rails aborted!
NameError: uninitialized constant ActionView::PathResolver

/Users/dan/.rbenv/versions/3.0.3/lib/ruby/gems/3.0.0/gems/versioncake-4.0.2/lib/versioncake.rb:27:in `<main>'
```

See https://github.com/rails/rails/pull/41969.

**Comment by trak3r** [comment]
I've take a few stabs at it this week but it seems the changes to `ActionView` are quite drastic. Each time you resolve one incompatibility you unveil a dozen more, and I'm not even trying to preserve backwards compatibility yet; I'm just hacking at it with a chainsaw.
If anybody else is trying to get this working please let me know and perhaps we can compare notes, or I can help you test and polish if you're making better progress.

**Comment by bwillis** [comment]
Thanks for opening up the issue and thanks @trak3r for taking a look at it! 

Phew, ok so it seems these Rails refactors to optimize the rendering seems to have made it quite a bit harder to extend our use case. I have been able to get it working in #110 but it's a ton of monkey patching. The removal of several of the constants made it necessary to override each method that was considering the template details so that we could inject version.

Due to the tight integration, I think this is too brittle at the moment to release at this point. I'll take some time and would appreciate any thoughts other folks have on #110 to see how we can simplify it. I've also been toying around with the idea of  refactoring Rails itself to make this integration simpler. Ideally it will have good side effects for the variant, locale, format & handler features, but of course this path would take more time though.

**Comment by trak3r** [comment]
I've tested this branch in two different sizable projects and it looks good. One of the projects hacks it significantly to use client IDs (resolved from subdomains) as the "versions" for serving different view templates and logic paths and it all seems fine. Would love to see an "official" release 😉  😉

**Comment by bwillis** [comment]
> to use client IDs (resolved from subdomains) as the "versions"

Oh, that's a neat use case.

> Would love to see an "official" release

Yeah let me see if I can comb through the MR again this weekend to see if I can restructure it to feel more confident in merging it.

**Comment by trak3r** [comment]
This is now the last gem holding up my Rails 7 upgrade 🙏🏻 🥺 😄 pretty please

**Comment by mladenoff** [comment]
How are we looking on #110 is there more work to be done to get it into a mergeable state?

**Comment by trak3r** [comment]
I'm getting this error after new release (4.1.0) and suspect it's related

`NameError: uninitialized constant ANY_HASH`

versioncake rails 7 pull request
https://github.com/bwillis/versioncake/pull/110/files

breaking code from rails 7
https://github.com/rails/rails/blob/main/actionview/lib/action_view/template_details.rb#L9

**Comment by trak3r** [comment]
UPDATE i see @bwillis originally fixed it here
https://github.com/bwillis/versioncake/pull/110/commits/3368d5b651e41f2b57475e0664a4bb659a3397e2

then seemingly re-broke it here 
https://github.com/bwillis/versioncake/pull/110/commits/76cbc3759d1a0a02755503ce8090074d095d659a#diff-59ef8b607e0a0acc2fa9a612f5cbb8ad4d4c4e3d51088cc5e16c45413e4758b0R61

**Comment by bwillis** [comment]
@trak3r thanks for checking, just fixed in #116, can you check against master? I'll push an update if you give the all clear.

**Comment by trak3r** [comment]
@bwillis 

> @trak3r thanks for checking, just fixed in #116, can you check against master? I'll push an update if you give the all clear.

`gem 'versioncake', git: 'https://github.com/bwillis/versioncake', branch: 'master'`

confirmed tests passing for me now. thanks.

**Comment by bwillis** [comment]
@trak3r thanks a lot for the quick turnaround and testing, I've released v4.1.1 with that fix, so I think we're good enough to call Rails 7 support closed. Please feel free to create any new issues if you see something! 🐛
